pkg/slo-controller/nodeslo/resource_strategy.go

```diff
@@ -107,65 +72,116 @@ func getResourceQoSSpec(node *corev1.Node, configMap *corev1.ConfigMap) (*slov1a
-	cfgStr, ok := configMap.Data[config.CPUBurstConfigKey]
+func caculateResourceThresholdCfgMerged(oldCfg config.ResourceThresholdCfg, configMap *corev1.ConfigMap) (config.ResourceThresholdCfg, error) {
+	cfgStr, ok := configMap.Data[config.ResourceThresholdConfigKey]
 	if !ok {
-		return mergedStrategy, nil
+		return DefaultSLOCfg().ThresholdCfgMerged, nil
 	}
 
-	cfg := config.CPUBurstCfg{}
-	if err := json.Unmarshal([]byte(cfgStr), &cfg); err != nil {
-		klog.Warningf("failed to unmarshal config %s, error: %v", config.CPUBurstConfigKey, err)
-		return nil, err
+	mergedCfg := config.ResourceThresholdCfg{}
+	if err := json.Unmarshal([]byte(cfgStr), &mergedCfg); err != nil {
+		klog.Errorf("failed to unmarshal config %s, err: %s", config.ResourceThresholdConfigKey, err)
+		return oldCfg, err
 	}
-	klog.V(5).Infof("get cpu burst config spec, origin string %v, cluster %+v, node %+v",
-		cfgStr, cfg.ClusterStrategy, cfg.NodeStrategies)
 
-	if cfg.ClusterStrategy != nil {
-		mergedStrategyInterface, err := util.MergeCfg(mergedStrategy, cfg.ClusterStrategy)
-		if err != nil {
-			klog.Warningf("merge cluster config for cpu burst failed, error %v", err)
-			return nil, err
+	// merge ClusterStrategy
+	clusterMerged := DefaultSLOCfg().ThresholdCfgMerged.ClusterStrategy.DeepCopy()
+	if mergedCfg.ClusterStrategy != nil {
+		mergedStrategyInterface, _ := util.MergeCfg(clusterMerged, mergedCfg.ClusterStrategy)
+		clusterMerged = mergedStrategyInterface.(*slov1alpha1.ResourceThresholdStrategy)
+	}
+	mergedCfg.ClusterStrategy = clusterMerged
+
+	for index, nodeStrategy := range mergedCfg.NodeStrategies {
+		// merge with clusterStrategy
+		clusterCfgCopy := mergedCfg.ClusterStrategy.DeepCopy()
+		if nodeStrategy.ResourceThresholdStrategy != nil {
+			mergedNodeStrategyInterface, _ := util.MergeCfg(clusterCfgCopy, nodeStrategy.ResourceThresholdStrategy)
+			mergedCfg.NodeStrategies[index].ResourceThresholdStrategy = mergedNodeStrategyInterface.(*slov1alpha1.ResourceThresholdStrategy)
+		} else {
+			mergedCfg.NodeStrategies[index].ResourceThresholdStrategy = clusterCfgCopy
 		}
-		mergedStrategy = mergedStrategyInterface.(*slov1alpha1.CPUBurstStrategy)
+
 	}
 
-	sort.Slice(cfg.NodeStrategies, func(i, j int) bool {
-		return cfg.NodeStrategies[i].NodeSelector.String() < cfg.NodeStrategies[j].NodeSelector.String()
-	})
+	return mergedCfg, nil
+}
 
-	nodeLabels := labels.Set(node.Labels)
-	for _, nodeStrategy := range cfg.NodeStrategies {
-		selector, err := metav1.LabelSelectorAsSelector(nodeStrategy.NodeSelector)
-		if err != nil {
-			klog.Errorf("failed to parse node selector %v, err: %v", nodeStrategy.NodeSelector, err)
-			continue
-		}
-		if !selector.Matches(nodeLabels) {
-			continue
-		}
-		if nodeStrategy.CPUBurstStrategy == nil {
-			continue
+func caculateResourceQoSCfgMerged(oldCfg config.ResourceQoSCfg, configMap *corev1.ConfigMap) (config.ResourceQoSCfg, error) {
+	cfgStr, ok := configMap.Data[config.ResourceQoSConfigKey]
```

Comment:
the default resource qos set by the controller should be empty. e.g. `return &slov1alpha1.ResourceQoSStrategy{}`

Reply:
DefaultSLOCfg().ResourceQoSCfgMerged is also  empty?

Reply:
Yes. Same as the former implementation, the final version of resourceQoS config derives from koordlet. In the slo-controller, we want a merged version of custom configurations (i.e. clusterStrategy in configMap merge with nodeStrategy), and the merge with default is not wanted.